### DISCUSSION TITLE
commit done wrong

### DISCUSSION THREAD

**Discussion by Saiipe** [community/community#65829] [Other Feature Feedback, Questions, & Ideas]
### Select Topic Area

Question

### Body

I made a wrong commit, directly from the github website, and I would like to know if there is any way to delete the commit, the commit itself is silly, but if it was an important code I would not know how to remove it, can someone help me ?

**Comment by LeandroLuna** [comment]
Sim, é possível excluir um commit de um repositório do GitHub.

Solução 1:

Se o commit estiver em uma branch que apenas você está trabalhando (por exemplo, uma branch de recurso), você pode fazer um push forçado da branch para o GitHub após fazer as alterações necessárias para remover o commit. Aqui estão os passos:

a. Localmente, use `git reset --hard HEAD~1` para remover o último commit da sua branch.

b. Em seguida, use `git push origin <nome-da-branch> --force` para fazer um push forçado da branch para o GitHub. Substitua `<nome-da-branch>` pelo nome real da sua branch.

Tenha extrema cautela com o push forçado, pois ele reescreve o histórico. Certifique-se de que você não está afetando uma branch compartilhada.

Solução 2:

Em vez de excluir um commit, você também pode considerar revertê-lo. Isso cria um novo commit que desfaz as alterações feitas no commit anterior, mantendo o histórico de commits intacto. Para fazer isso, você pode usar o seguinte comando:

   ```bash
   git revert <hash-do-commit>
   ```
Substitua `<hash-do-commit>` pela hash real do commit que você deseja reverter. Esta abordagem é mais segura para branches compartilhadas.

Por favor, caso a solução tenha lhe servido, peço que aceite como resposta final da postagem.
Abraços!

**Reply by Saiipe** [reply]
a questão é que eu fiz o commit diretamente do site do github, eu queria saber se tem como apagar pelo próprio site.

**Comment by GhostOf0days** [comment]
1. Download [Git](https://git-scm.com/downloads).
2. Make sure your local repo is a clone.
3. Open a terminal in your code editor.
4. Enter 'git checkout main'
5. Enter 'git reset --hard HEAD~1.'
6. Enter 'git push --f."

**Comment by seyLu** [comment]
#### Hi  @Saiipe 

#### This is not possible in Github web and for good reasons.

#### Generally, it is frowned upon to change git history, especially if you have existing users in your project. In fact, to protect against this, you can set branch protection rules in Github to disallow force pushes.

- read more on:
  * [Practical consequences of rewriting git history](https://stackoverflow.com/questions/1491001/what-are-the-practical-consequences-of-rewriting-git-history)
  * [About protected branches](https://docs.github.com/en/repositories/configuring-branches-and-merges-in-your-repository/managing-protected-branches/about-protected-branches)
 
#### What you can do though, instead of deleting the commit, is to go back in git history, but keep the commit. Something like:

```mermaid
gitGraph
 commit tag: "Initial Commit"
 commit
 commit tag: "Good Commit #1"
 commit
 commit tag: "Good Commit #2"
 commit type: HIGHLIGHT
 commit tag: "Bad Commits" type: HIGHLIGHT
 commit type: HIGHLIGHT
 commit tag: "Good Commit #2" type: REVERSE
```

#### To do this in CLI (as answered by  @LeandroLuna):

```bash
# list commit history
# <commit_hash> <commit_message>
git log --oneline


# revert back to an older commit hash
git revert <commit_hash>
```

#### As far as reverting/deleting commits in Github web, currently, I do not think it is possible.
